lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php

```diff
@@ -29,44 +40,44 @@ abstract class AbstractMySQLDriver implements Driver, ExceptionConverterDriver,
      * @link https://dev.mysql.com/doc/refman/8.0/en/client-error-reference.html
      * @link https://dev.mysql.com/doc/refman/8.0/en/server-error-reference.html
      */
-    public function convertException($message, DriverException $exception)
+    public function convertException($message, Exception $exception)
```

Comment:
Small BC-break here: classes extending this one will narrow the type, which breaks the LSP. I'm not saying it has to be a blocker, but maybe in these 103 files there are similar BC-breaks that are more serious, so I raise this in case you did not have that in mind.

Slightly off-topic, have we ever considered using roave's BC-checker on this package?

Reply:
> Small BC-break here: classes extending this one will narrow the type, which breaks the LSP. I'm not saying it has to be a blocker, but maybe in these 103 files there are similar BC-breaks that are more serious, so I raise this in case you did not have that in mind.

This is a valid point, I didn't have that in mind. To be absolutely on the safe side, we may only introduce the new classes but not change any method signatures that accept them (we could change the return types but we probably don't have many of them in `2.11.x`).

> Slightly off-topic, have we ever considered using roave's BC-checker on this package?

It might be a good time to give it a try.

Reply:
Is there still a point in introducing the new classes? Can they be used anywhere?

[I have a solution to that kind of issue](https://dev.to/greg0ire/how-to-deprecate-a-type-in-php-48cf), and we went with it when migrating the `Doctrine\Common\Persistence` namespace to just `Doctrine\Persistence`, and it was painful (Andreas wrote several blog posts about it). We vowed never to do that again, so I'm not sure how this is going to pan out. Maybe you can find a subset of this PR that consists in classes never used in method signatures that can be overriden outside of the DBAL?

This is why I tend to `final` everything…

Reply:
Most of these classes can be finalized (https://github.com/doctrine/dbal/pull/3590, https://github.com/doctrine/dbal/issues/3818) but not all of them. We indeed can/should mark all the new ones that can be marked final right now instead of waiting.